### PR TITLE
Initial Web application with Carbon and Grafana Visualizations.

### PR DESCRIPTION
**Pull request by bmath10**
This PR includes both the initial Grafana visulizations and the initial set-up of the web application using Carbon Components based off of Mike's designs.

**TO ADDRESS PROTRACTOR ISSUES**
- I've switched the project to yarn from npm. Instructions to install yarn can be found here: https://yarnpkg.com/en/docs/install#mac-stable . This shouldn't change anything besides all npm commands should be run with yarn (they'll still work with npm though).
- To make sure you're downloading from scratch, `rm -rf node_modules/` in the root directory
- `yarn install` in the root directory
- Then make sure to run `yarn run pre-e2e`, following the instructions below if you get stuck, and finally the end-to-end tests with `yarn run e2e`

**NOTES**
- You have to run `yarn install` in the root directory (the one with the `package.json` file) before running/testing.
- Adding the carbon component styling to the .scss files has increased compiling time significantly, so don't be surprised when the `ng` commands take around a minute.
- Running the `yarn run e2e` command (end-to-end testing) will most likely require redownloading the chromedriver for selenium, and instructions can be found here: https://gitlab.mitre.org/lantern/nms-backend/wikis/Angular-Set-Up
- Reminder that testing instructions can be found here: https://github.com/onc-healthit/lantern-front-end/wiki/Running-the-Project

**CHECKLIST**

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:** @bmath10 
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-37 and https://oncprojectracking.healthit.gov/support/browse/LANTERN-51
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer: @edeyoung**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge is complete.

### CODE REVIEW

**Review by mrnosal — CHANGES_REQUESTED** [review]
The protractor installed chromedriver doesn't work with Chrome 78, just 76

File: e2e/src/app.e2e-spec.ts [hunk 1]
```diff
@@ -8,9 +8,31 @@ describe('workspace-project App', () => {
...
+    element(by.className('about-link')).click();
+    const subheader = $('.sub-header').$('h1');
+    expect(subheader.getText()).toEqual('About Lantern');
+  });
+
+  it('should navigate to Home page', () => {
+    page.navigateTo();
+    element(by.className('bx--header__name')).click();
+    const homeBody = $$('.home-body h5').first();
+    expect(homeBody.getText()).toEqual('Endpoint Metrics');
```
Comment (edeyoung) [inline e2e/src/app.e2e-spec.ts:38]: should there also be a test that says the default page is the home page (and the homebody equals 'Endpoint Metrics')? Similar to the 'Lantern' test above.

File: src/app/components/endpoint-page/endpoint-page.component.scss [hunk 2]
```diff
@@ -0,0 +1,9 @@
+@import '_variables.scss';
+
+.format-body {
+    padding: $page-padding;
+
+    p {
+        padding-bottom: $base-padding;
+    }
+}
```
Comment (edeyoung) [inline src/app/components/endpoint-page/endpoint-page.component.scss:9]: is it standard to have each component's style file repeat styling? Can style files inherit from a more global file?

File: src/app/components/home-page/home-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-home-page',
+  templateUrl: './home-page.component.html',
+  styleUrls: ['./home-page.component.scss']
+})
+export class HomePageComponent implements OnInit {
+
+  public dashboardURL = 'http://ec2-3-228-127-99.compute-1.amazonaws.com/d-solo/ZEcnMHoWk/lantern?orgId=1&refresh=10s&theme=light';
```
Comment (edeyoung) [inline src/app/components/home-page/home-page.component.ts:10]: we talked about figuring out when it made sense for the front end to have a config file. This seems like something that would make sense in a config file. thoughts?
If you agree, can you add a "TODO: config file" comment and move the frontend config file task to the top of the backlog?

File: src/app/components/home-page/home-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  public totalEndpoints = this.dashboardURL + '&panelId=14';
+  public success = this.dashboardURL + '&panelId=12';
+  public unsuccessful = this.dashboardURL + '&panelId=16';
+  public unresponsive = this.dashboardURL + '&panelId=18';
+  public newEndpoints = this.dashboardURL + '&panelId=20';
+  public responseTimeRange = this.dashboardURL + '&panelId=22';
+  public currentResponseTime = this.dashboardURL + '&panelId=8';
```
Comment (edeyoung) [inline src/app/components/home-page/home-page.component.ts:18]: does it make sense to include tests related to these grafana vizs?

File: src/app/components/shared/page-footer/page-footer.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(PageFooterComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
Comment (edeyoung) [inline src/app/components/shared/page-footer/page-footer.component.spec.ts:28]: does it make sense to add tests re the content of the footer? or not so much because that's in the *.html rather than the actual logic?

File: src/app/components/shared/page-header/page-header.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(PageHeaderComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (edeyoung) [inline src/app/components/shared/page-header/page-header.component.spec.ts:31]: similar question as for footer re testing content.

**Review by edeyoung — COMMENTED** [review]
Had a few questions.

**Comment by bmath10** [comment]
> The protractor installed chromedriver doesn't work with Chrome 78, just 76

I updated protractor in the package.json and that did seem to fix it for me (unclear why but this does seem to be a known issue). If you pull the latest, run `npm install` and try just running the `npm run pre-e2e` command, everything should work? If not, let me know cause then I maybe accidentally fixed it yesterday trying random things without realizing.

File: e2e/src/app.e2e-spec.ts (hunk 1 above)
Comment (bmath10) [reply e2e/src/app.e2e-spec.ts:38]: I think that's a good point, I'll add another test to cover that.

File: src/app/components/endpoint-page/endpoint-page.component.scss (hunk 2 above)
Comment (bmath10) [reply src/app/components/endpoint-page/endpoint-page.component.scss:9]: To answer the questions that you actually asked, it is not uncommon for components to have repeat styling (though it's not good practice), and style files can inherit from a more global file.

**Tl;dr**: The common styling here should be extracted to a separate file that both of these styling files import.

**Too long and going to read anyways**: 
1.  In the olden web application days, where you were just importing all css files to your main index.html or whatever, all classname/tag/etc. styling applied to everything and the styling that had priority was from the stylesheet that was last imported in the list. This horrible system lead to Angular using styling encapsulation to make it a lot easier to style your html and figure out why your html was being styled a specific way. This doesn't really relate directly to your question but it might give some helpful context.
2. Sometimes, especially in more complicated projects, you'll see a lot of repeat styling copy-pasted. This is usually because as an app is being developed styling tends to change a lot in dramatic ways, so as good practice you might start extracting common styling into separate files and then later no components are actually using those class names or styling so there's lot of little remnants everywhere that people don't bother to clean up (obviously not good practice, but it is incredibly common). 
3. People tend to avoid putting styling in the global files like `styles.scss` at the root level because the component style files are encapsulated, whereas `styles.scss` applies to all components. So if you put in that root level file that all `h1` tags have padding 10px, and then later you have an h1 tag in a component that you don't want padding for, that can lead to the nested overwriting that I mentioned in the first point. The component encapsulation also helps in the case where you're reusing class names for convenience, like `home-body`, in different components but you don't want them to share styling (e.g. one file has `.home-body { padding: 10px }` and another has `.home-body { background-color: blue }` which would have been a problem in the older web application projects. 
4. We can create general styling files and important them into component styling files so that the common styling can apply to both files, which is what I think makes sense here. I did consider briefly that it could make sense to make a generic page component that the about page and endpoint page could inherit from, but since these pages will ultimately end up very different, just having a common styling sheet that they both import makes the most sense.

File: src/app/components/home-page/home-page.component.ts (hunk 3 above)
Comment (bmath10) [reply src/app/components/home-page/home-page.component.ts:10]: I agree, I think this makes sense since this has already caused issues!

**Comment by bmath10** [comment]
TODO: config file

File: src/app/components/endpoint-page/endpoint-page.component.scss (hunk 2 above)
Comment (edeyoung) [reply src/app/components/endpoint-page/endpoint-page.component.scss:9]: good context, thank you! sounds like long term, yes, shared style file. Short term - things are in flux so figure out what to extract out at a later time. let me know if didn't capture that correctly.

File: src/app/components/home-page/home-page.component.ts (hunk 4 above)
Comment (bmath10) [reply src/app/components/home-page/home-page.component.ts:18]: I think so, the only tests I can think of that make sense to me are making sure they get loaded into the html properly and that they reference valid panel ids. Was there anything else you had in mind?

File: src/app/components/shared/page-footer/page-footer.component.spec.ts (hunk 5 above)
Comment (bmath10) [reply src/app/components/shared/page-footer/page-footer.component.spec.ts:28]: In general I don't think it's worth it since it's not loading anything from the ts, it's just straight html.

**Comment by mrnosal** [comment]
It is still installing chromedriver 76

File: src/app/components/home-page/home-page.component.ts (hunk 4 above)
Comment (bmath10) [reply src/app/components/home-page/home-page.component.ts:18]: I wasn't able to write a test to see if they have valid panel ids because you apparently can't capture iframe errors since it's a cross-domain request? It seems like a good idea to check if these urls return errors but i didn't see a great way to do that, at least at first glance.

**Comment by bmath10** [comment]
currently waiting to push any yarn stuff until I hear from @edeyoung that the right chromedriver either does or does not download!

**Review by mrnosal — APPROVED** [review]
yarn is working correctly with chrome 78 driver

**Review by edeyoung — APPROVED** [review]
